pkg/initializer/webhook_initializer.go

```diff
@@ -0,0 +1,156 @@
+//  - create or refresh Secret and CSR for ca cert
+//  - update webhook server
+//  - update validationAdmissionConfiguration
+func (initializer *Initializer) webhookResourceIntializer(podName, namespace string, days int) error {
+
+	if !WebhookEnabled {
+		return nil
+	}
+	klog.Info("initializer start to generate resources for webhook")
+
+	err := GenerateSecretAndCSR(AdmissionWebhookName, namespace, days)
+	if err != nil {
+		return err
+	}
+	klog.Infof("success to apply CA cert for service[%s/%s]", namespace, AdmissionWebhookName)
+
+	secret, err := initializer.kubeCli.CoreV1().Secrets(namespace).Get(SecretNameForServiceCert(AdmissionWebhookName), metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	csrName := fmt.Sprintf("%s.%s-csr", AdmissionWebhookName, namespace)
+
+	csr, err := initializer.kubeCli.CertificatesV1beta1().CertificateSigningRequests().Get(csrName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	err = initializer.setOwnerReferences(podName, namespace, secret, csr)
+	if err != nil {
+		return err
+	}
+
+	_, err = initializer.kubeCli.CoreV1().Secrets(namespace).Update(secret)
+	if err != nil {
+		return err
+	}
+	_, err = initializer.kubeCli.CertificatesV1beta1().CertificateSigningRequests().Update(csr)
+	if err != nil {
+		return err
+	}
+
+	err = initializer.updateWebhookServer(namespace, secret)
+	if err != nil {
+		klog.Errorf("failed to update webhook server[%s/%s],%v", namespace, AdmissionWebhookName, err)
+		return err
+	}
+	klog.Infof("success to update webhook server[%s/%s]", namespace, AdmissionWebhookName)
+
+	err = initializer.updateValidationAdmissionConfiguration(secret)
+	if err != nil {
+		klog.Errorf("failed to update validation admission config[%s/%s],%v", namespace, VACNameForService(namespace, AdmissionWebhookName), err)
+		return err
+	}
+	klog.Infof("success to update validation admission config[%s/%s]", VACNameForService(namespace, AdmissionWebhookName), err)
+	return nil
+}
+
+// update tidb-operator validation webhook server
+func (initializer *Initializer) updateValidationAdmissionConfiguration(secret *core.Secret) error {
+
+	conf, err := initializer.kubeCli.AdmissionregistrationV1beta1().ValidatingWebhookConfigurations().Get(AdmissionWebhookName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	f := admissionregistration.Fail
+	for id, webhook := range conf.Webhooks {
+		webhook.ClientConfig.CABundle = secret.Data["cert.pem"]
+		webhook.FailurePolicy = &f
+		conf.Webhooks[id] = webhook
+	}
+	_, err = initializer.kubeCli.AdmissionregistrationV1beta1().ValidatingWebhookConfigurations().Update(conf)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// update Webhook server to make sure the latest secret would be used.
+func (initializer *Initializer) updateWebhookServer(namespace string, secret *core.Secret) error {
+
+	server, err := initializer.kubeCli.ExtensionsV1beta1().Deployments(namespace).Get(AdmissionWebhookName, metav1.GetOptions{})
```

Comment:
This hard-coded name an implicit dependency between the initializer binary and the helm chart.

IMHO, we should provide a command line argument to specify the deployment name, and coordinate the names in the helm chart.

Reply:
fix